File: problem_21.rb
class AmicableNumbersSum

  private

  attr_accessor :up_to, :amicable_numbers
  attr_writer :sum

  def initialize(up_to)
    self.up_to = up_to
    self.amicable_numbers = find_amicable
    self.sum = amicable_numbers.sum
  end

  def find_amicable
    (1..up_to).to_a.select do |number|
      sum_of_divisors(sum_of_divisors(number)) == number &&
        sum_of_divisors(number) != number
    end
  end

  def sum_of_divisors(number)
    (1...number).select { |candidate| (number % candidate).zero? }.sum
  end

  public

  attr_reader :sum, :amicable_numbers

end

if __FILE__ == $PROGRAM_NAME
  a =  AmicableNumbersSum.new(10000)
  puts "The amicable numbers up to 10000 are: %s." % a.amicable_numbers.to_s
  puts "Their sum is %s." % a.sum
end

File: problems/problem_21.py
def main(up_to):
    return [find_amicable(up_to), sum(find_amicable(up_to))]

def find_amicable(up_to):
    return [x for x in range(1, up_to + 1) if sum_of_divisors(x) != x and
            sum_of_divisors(sum_of_divisors(x)) == x]


def sum_of_divisors(number):
    return sum([x for x in range(1, number) if number % x == 0])

if __name__ == "__main__":
    a = main(10000)
    b, c = a[0], a[1]
    print("The amicable numbers up to 10000 are: " + str(b))
    print("Their sum is " + str(c) + ".")
